File: client/src/main/scala/stasis/client/ops/recovery/Recovery.scala
package stasis.client.ops.recovery

import java.nio.file.FileSystem
import java.nio.file.FileSystems
import java.nio.file.Path
import java.time.Instant

import scala.concurrent.ExecutionContext
import scala.concurrent.Future
import scala.util.Failure
import scala.util.Success
import scala.util.matching.Regex

import org.apache.pekko.Done
import org.apache.pekko.NotUsed
import org.apache.pekko.actor.typed.ActorSystem
import org.apache.pekko.stream._
import org.apache.pekko.stream.scaladsl.Sink
import org.apache.pekko.stream.scaladsl.Source

import stasis.client.collection.RecoveryCollector
import stasis.client.collection.RecoveryMetadataCollector
import stasis.client.encryption.secrets.DeviceSecret
import stasis.client.model.DatasetMetadata
import stasis.client.model.TargetEntity
import stasis.client.ops.ParallelismConfig
import stasis.client.ops.exceptions.EntityProcessingFailure
import stasis.client.ops.exceptions.OperationStopped
import stasis.client.ops.recovery.stages.EntityCollection
import stasis.client.ops.recovery.stages.EntityProcessing
import stasis.client.ops.recovery.stages.MetadataApplication
import stasis.client.tracking.RecoveryTracker
import stasis.shared.model.datasets.DatasetDefinition
import stasis.shared.model.datasets.DatasetEntry
import stasis.shared.ops.Operation

class Recovery(
  descriptor: Recovery.Descriptor
)(implicit system: ActorSystem[Nothing], parallelism: ParallelismConfig, providers: Providers)
    extends Operation { parent =>
  import Recovery._

  private implicit val mat: Materializer = SystemMaterializer(system).materializer

  private val supervision: Supervision.Decider = {
    case e: EntityProcessingFailure =>
      system.log.error(
        "Recovery stream encountered failure while processing entity [{}]: [{} - {}]; resuming",
        e.entity,
        e.cause.getClass.getSimpleName,
        e.cause.getMessage
      )
      providers.track.failureEncountered(e.entity, failure = e.cause)
      Supervision.Resume

    case e =>
      system.log.error("Recovery stream encountered failure: [{} - {}]; resuming", e.getClass.getSimpleName, e.getMessage)
      providers.track.failureEncountered(failure = e)
      Supervision.Resume
  }

  private implicit val ec: ExecutionContext = system.executionContext

  override implicit val id: Operation.Id = Operation.generateId()

  override val `type`: Operation.Type = Operation.Type.Recovery

  private val collector: RecoveryCollector = parent.descriptor.toRecoveryCollector()

  private implicit val killSwitch: SharedKillSwitch = KillSwitches.shared("recovery-kill-switch")

  private val stream: Source[Done, NotUsed] =
    stages.entityCollection
      .via(stages.entityProcessing)
      .via(stages.metadataApplication)
      .via(killSwitch.flow)
      .withAttributes(ActorAttributes.supervisionStrategy(supervision))

  override def start(): Future[Done] =
    stream
      .runWith(Sink.ignore)
      .trackWith(providers.track)

  override def stop(): Unit =
    killSwitch.abort(OperationStopped(s"Operation [${id.toString}] stopped by user"))

  private object stages extends EntityCollection with EntityProcessing with MetadataApplication {
    override protected lazy val deviceSecret: DeviceSecret = parent.descriptor.deviceSecret
    override protected lazy val providers: Providers = parent.providers
    override protected lazy val collector: RecoveryCollector = parent.collector
    override protected lazy val parallelism: ParallelismConfig = parent.parallelism
    override implicit protected lazy val mat: Materializer = parent.mat
    override implicit protected lazy val ec: ExecutionContext = parent.system.executionContext
  }
}

object Recovery {
  def apply(
    descriptor: Recovery.Descriptor
  )(implicit system: ActorSystem[Nothing], parallelism: ParallelismConfig, providers: Providers): Recovery =
    new Recovery(descriptor)

  final case class Descriptor(
    targetMetadata: DatasetMetadata,
    query: Option[PathQuery],
    destination: Option[Destination],
    deviceSecret: DeviceSecret
  ) {
    def toRecoveryCollector()(implicit
      ec: ExecutionContext,
      mat: Materializer,
      parallelism: ParallelismConfig,
      providers: Providers
    ): RecoveryCollector =
      new RecoveryCollector.Default(
        targetMetadata = targetMetadata,
        keep = (entity, _) => query.forall(_.matches(entity.toAbsolutePath)),
        destination = destination.toTargetEntityDestination,
        metadataCollector = RecoveryMetadataCollector.Default(checksum = providers.checksum),
        api = providers.clients.api
      )
  }

  object Descriptor {
    sealed trait Collector
    object Collector {
      final case class WithDefinition(definition: DatasetDefinition.Id, until: Option[Instant]) extends Collector
      final case class WithEntry(entry: DatasetEntry.Id) extends Collector
    }

    def apply(
      query: Option[PathQuery],
      destination: Option[Destination],
      collector: Descriptor.Collector,
      deviceSecret: DeviceSecret
    )(implicit ec: ExecutionContext, providers: Providers): Future[Descriptor] =
      for {
        entry <- collector match {
          case Collector.WithDefinition(definition, until) =>
            providers.clients.api.latestEntry(definition, until).flatMap {
              case Some(entry) =>
                Future.successful(entry)

              case None =>
                Future.failed(
                  new IllegalStateException(
                    s"Expected dataset entry for definition [${definition.toString}] but none was found"
                  )
                )
            }

          case Collector.WithEntry(entry) =>
            providers.clients.api.datasetEntry(entry)
        }
        metadata <- providers.clients.api.datasetMetadata(entry)
      } yield {
        Descriptor(
          targetMetadata = metadata,
          query = query,
          destination = destination,
          deviceSecret = deviceSecret
        )
      }
  }

  sealed trait PathQuery {
    def matches(path: Path): Boolean
  }

  object PathQuery {
    def apply(query: String): PathQuery =
      if (query.contains("/")) {
        ForAbsolutePath(query = new Regex(query))
      } else {
        ForFileName(query = new Regex(query))
      }

    final case class ForAbsolutePath(query: Regex) extends PathQuery {
      override def matches(path: Path): Boolean =
        query.pattern.matcher(path.toAbsolutePath.toString).find()
    }

    final case class ForFileName(query: Regex) extends PathQuery {
      override def matches(path: Path): Boolean =
        query.pattern.matcher(path.getFileName.toString).find()
    }
  }

  final case class Destination(
    path: String,
    keepStructure: Boolean,
    filesystem: FileSystem
  )

  object Destination {
    def apply(path: String, keepStructure: Boolean): Destination =
      Destination(path = path, keepStructure = keepStructure, filesystem = FileSystems.getDefault)
  }

  implicit class RecoveryToTargetEntityDestination(destination: Option[Destination]) {
    def toTargetEntityDestination: TargetEntity.Destination =
      destination match {
        case Some(destination) =>
          TargetEntity.Destination.Directory(
            path = destination.filesystem.getPath(destination.path),
            keepDefaultStructure = destination.keepStructure
          )

        case None =>
          TargetEntity.Destination.Default
      }
  }

  implicit class TrackedOperation(operation: Future[Done]) {
    def trackWith(track: RecoveryTracker)(implicit id: Operation.Id, ec: ExecutionContext): Future[Done] = {
      operation.onComplete {
        case Success(_) =>
          track.completed()

        case Failure(_: OperationStopped) =>
          () // do nothing

        case Failure(e: EntityProcessingFailure) =>
          track.failureEncountered(entity = e.entity, failure = e.cause)

        case Failure(e) =>
          track.failureEncountered(failure = e)
      }

      operation
    }
  }
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/recovery/Recovery.kt
package stasis.client_android.lib.ops.recovery

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.plus
import stasis.client_android.lib.encryption.secrets.DeviceSecret
import stasis.client_android.lib.model.DatasetMetadata
import stasis.client_android.lib.model.FilesystemMetadata
import stasis.client_android.lib.model.TargetEntity
import stasis.client_android.lib.model.server.datasets.DatasetDefinitionId
import stasis.client_android.lib.model.server.datasets.DatasetEntryId
import stasis.client_android.lib.ops.Operation
import stasis.client_android.lib.ops.OperationId
import stasis.client_android.lib.ops.recovery.Recovery.Destination.Companion.toTargetEntityDestination
import stasis.client_android.lib.ops.recovery.stages.EntityCollection
import stasis.client_android.lib.ops.recovery.stages.EntityProcessing
import stasis.client_android.lib.ops.recovery.stages.MetadataApplication
import stasis.client_android.lib.utils.Try
import stasis.client_android.lib.utils.Try.Companion.flatMap
import stasis.client_android.lib.utils.Try.Companion.map
import stasis.client_android.lib.utils.Try.Failure
import stasis.client_android.lib.utils.Try.Success
import java.nio.file.FileSystem
import java.time.Instant
import java.util.concurrent.atomic.AtomicReference

class Recovery(
    private val descriptor: Descriptor,
    private val providers: Providers
) : Operation {
    override val id: OperationId = Operation.generateId()

    override fun type(): Operation.Type = Operation.Type.Recovery

    override fun start(withScope: CoroutineScope, f: (Throwable?) -> Unit) {
        require(jobRef.get() == null) { "Recovery [$id] already started" }

        val supervisor = SupervisorJob()

        val job = (withScope + supervisor).async {
            try {
                providers.track.started(operation = id)

                stages.entityCollection(id, descriptor.filesystem)
                    .let { flow -> stages.entityProcessing(id, flow) }
                    .let { flow -> stages.metadataApplication(id, flow) }
                    .collect()

                providers.track.completed(id)
            } catch (e: Throwable) {
                providers.track.failureEncountered(id, e)
                providers.analytics.recordFailure(e)
                throw e
            }
        }

        job.invokeOnCompletion { e ->
            supervisor.cancel()
            f(e)
        }

        jobRef.set(job)
    }

    override fun stop() {
        val job = jobRef.get()
        require(job != null) { "Recovery [$id] not started" }

        job.cancel(cause = CancellationException("Cancelled by user"))
    }

    private val jobRef: AtomicReference<Job> = AtomicReference()

    private val stages = object : EntityCollection, EntityProcessing, MetadataApplication {
        override val deviceSecret: DeviceSecret = descriptor.deviceSecret
        override val providers: Providers = this@Recovery.providers
        override val targetMetadata: DatasetMetadata = descriptor.targetMetadata
        override val keep: (String, FilesystemMetadata.EntityState) -> Boolean =
            { entity, _ ->
                (descriptor.entities?.contains(entity) ?: true) &&
                    RecoverySourceKind.forEntity(entity) in descriptor.sources
            }
        override val destination: TargetEntity.Destination =
            descriptor.destination.toTargetEntityDestination(descriptor.filesystem)
    }

    data class Descriptor(
        val targetMetadata: DatasetMetadata,
        val entities: Set<String>?,
        val sources: Set<RecoverySourceKind>,
        val destination: Destination?,
        val deviceSecret: DeviceSecret,
        val filesystem: FileSystem
    ) {
        sealed class Collector {
            data class WithDefinition(val definition: DatasetDefinitionId, val until: Instant?) :
                Collector()

            data class WithEntry(val entry: DatasetEntryId) : Collector()
        }

        companion object {
            suspend operator fun invoke(
                entities: Set<String>?,
                sources: Set<RecoverySourceKind>,
                destination: Destination?,
                collector: Collector,
                deviceSecret: DeviceSecret,
                providers: Providers,
                filesystem: FileSystem
            ): Try<Descriptor> {
                val entry = when (collector) {
                    is Collector.WithDefinition -> providers.clients.api.latestEntry(
                        definition = collector.definition,
                        until = collector.until
                    ).flatMap { entry ->
                        entry
                            ?.let { Success(it) }
                            ?: Failure(
                                IllegalStateException(
                                    "Expected dataset entry for definition [${collector.definition}] but none was found"
                                )
                            )
                    }

                    is Collector.WithEntry -> providers.clients.api.datasetEntry(collector.entry)
                }

                return entry
                    .flatMap { providers.clients.api.datasetMetadata(it) }
                    .map { metadata ->
                        Descriptor(
                            targetMetadata = metadata,
                            entities = entities,
                            sources = sources,
                            destination = destination,
                            deviceSecret = deviceSecret,
                            filesystem = filesystem
                        )
                    }
            }
        }
    }

    data class Destination(
        val path: String,
        val keepStructure: Boolean,
        val preserveExisting: Boolean,
    ) {
        companion object {
            fun Destination?.toTargetEntityDestination(filesystem: FileSystem): TargetEntity.Destination =
                when (this) {
                    null -> TargetEntity.Destination.Default
                    else -> TargetEntity.Destination.Directory(
                        path = filesystem.getPath(path),
                        keepDefaultStructure = keepStructure,
                        preserveExisting = preserveExisting
                    )
                }
        }
    }
}
